hoodie-client/src/main/java/com/uber/hoodie/func/BufferedIterator.java

```diff
@@ -0,0 +1,224 @@
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Used for buffering input records. Buffer limit is controlled by {@link #bufferMemoryLimit}. It internally samples
+ * every {@link #RECORD_SAMPLING_RATE}th record and adjusts number of records in buffer accordingly. This is done to
+ * ensure that we don't OOM.
+ */
+public class BufferedIterator<K extends HoodieRecordPayload, T extends HoodieRecord<K>> implements Iterator<T> {
+
+  private static Logger logger = LogManager.getLogger(BufferedIterator.class);
+  private static final int RECORD_SAMPLING_RATE = 64;
+  private static final int RECORD_CACHING_LIMIT = 128 * 1024;
+
+  /**
+   * We will be using record size to determine how many records we should cache and will change permits accordingly.
+   */
+  private final Semaphore rateLimiter = new Semaphore(1);
+  private int currentRateLimit = 1;
+  private final LinkedBlockingQueue<BufferedIteratorRecord<T>> buffer = new LinkedBlockingQueue<>();
+  private final long bufferMemoryLimit;
+  private final Schema schema;
+
+  private long avgRecordSizeOfSampledRecords = 0;
+  private long numOfSampledRecords = 0;
+  private final Iterator<T> internalIterator;
+
+  private final AtomicReference<Exception> hasFailed = new AtomicReference(null);
+  private final AtomicBoolean isDone = new AtomicBoolean(false);
+  private T nextRecord;
+
+  private final AtomicLong readCounter = new AtomicLong(0);
+
+  private final AtomicLong samplingRecordCounter = new AtomicLong(-1);
+
+  public BufferedIterator(final Iterator<T> iterator, HoodieWriteConfig config) {
+    this.internalIterator = iterator;
+    this.bufferMemoryLimit = config.getInsertWriteBufferLimit();
+    this.schema = HoodieIOHandle.createHoodieWriteSchema(config);
+  }
+
+  private void adjustBufferSize(final T record) throws InterruptedException {
+    if (this.samplingRecordCounter.incrementAndGet() % RECORD_SAMPLING_RATE != 0) {
+      return;
+    }
+    final long recordSize = SizeEstimator.estimate(record);
+    final long newAvgRecordSize =
+      Math.max(1, (avgRecordSizeOfSampledRecords * numOfSampledRecords + recordSize) / (numOfSampledRecords + 1));
+    final int newRateLimit =
+      (int) Math.min(RECORD_CACHING_LIMIT, Math.max(1, this.bufferMemoryLimit / newAvgRecordSize));
+
+    if (newRateLimit > currentRateLimit) {
+      rateLimiter.release(newRateLimit - currentRateLimit);
+    } else if (newRateLimit < currentRateLimit) {
+      rateLimiter.acquire(currentRateLimit - newRateLimit);
+    }
+    currentRateLimit = newRateLimit;
+    avgRecordSizeOfSampledRecords = newAvgRecordSize;
+    numOfSampledRecords++;
+  }
+
+  private void insertRecord(T t) throws Exception {
+    adjustBufferSize(t);
+    rateLimiter.acquire();
```

Comment:
is the permit record based or byte based? in adjustBufferSize, we seem to be acquiring/releasing as much permits as the buffer needs to shrink/grow in bytes.. but here, in `insertRecord` and as well as in `readNextRecord`, each permit seems to imply a record?  

Side note: if we are just sampling, what value do we use to reduce the buffer size by, once we picked an item off.?

Reply:
permit is based on record count (and not on bytes). in "adjustBufferSizeIfNeeded" we try to acquire/release multiple permits based on whether new limit has increased or decreased.